Create material points within elements fixes #14, #18, #19

cw646 opened the pull request:
New pull request, removes Eigen files.

cw646 committed:
:construction: Working on issue #18

cw646 committed:
:construction: Working on issue #14, #18, #19
Need further investigation into stress issues, stress currently set @ 0.

cw646 committed:
:construction: Working on issue #14, #18, #19
Removal of Eigen folder, include changes to cmake to support docker Eigen.

Commit:
:wrench: fix issue #20 for stress

Commit:
:art: add travis

Commit:
:art: improve issue #20, for store elements

cw646 committed:
:construction: Working on issue #14, #18, #19
Inclusion of loop for storing coordinates, solved stress' not working with MPM.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/cb-geo/mpm-point-generator?pullRequest=21) <br/>All committers have signed the CLA.

kks32 commented on include/gmsh.h:
```diff
@@ -1,15 +1,18 @@
 #ifndef MPM_POINT_GEN_GMSH_H_
 #define MPM_POINT_GEN_GMSH_H_
 
+#include "mesh.h"
```
These local include files should be at the end after `Eigen` includes

kks32 commented on include/gmsh.h:
```diff
@@ -21,17 +24,28 @@ class GMSH : public Mesh {
-  //! Return a vector of elements
-  std::vector<std::shared_ptr<Point<3>>> elements() const { return elements_; }
+  //! Return a map of element id & vertices id
+  std::map<double, std::array<double, 4>> elements() const { return elements_; }
```
IDs should be unsigned or int.

kks32 replied:
Is the number `4` fixed for all elements?

kks32 commented on include/gmsh.h:
```diff
@@ -1,15 +1,18 @@
+#include <Eigen/Dense>
 
 //! \brief Generate Material Points from GMSH file
 class GMSH : public Mesh {
```
Make class templatise with dimension. This avoids defining dimension elsewhere within GMSH

kks32 commented on include/gmsh.h:
```diff
@@ -40,15 +54,21 @@ class GMSH : public Mesh {
-
-  //! Vector of elements
-  std::vector<std::shared_ptr<Point<3>>> elements_;
+  std::vector<std::shared_ptr<Point<3>>> materialpoints_;
```
This should be from the template argument of dimension `Tdim`

kks32 commented on include/gmsh.h:
```diff
@@ -40,15 +54,21 @@ class GMSH : public Mesh {
+  //! Map to store element ID and vertices ID
+  // 4 = element vertices
+  // 12 = total number of coordinate values for vertices
+  std::map<double, std::array<double, 4>> elements_;
```
Change it to a map of unsigned to a vector of unsigned.

kks32 commented on include/gmsh.h:
```diff
@@ -40,15 +54,21 @@ class GMSH : public Mesh {
+  // 4 = element vertices
+  // 12 = total number of coordinate values for vertices
+  std::map<double, std::array<double, 4>> elements_;
+  std::map<double, std::array<double, 12>> elementcoordinates_;
```
These coordinates are vertices coords, right?

kks32 commented on include/gmsh.tcc:
```diff
@@ -66,9 +64,13 @@ void GMSH::read_elements(const std::string& filename) {
 
   const unsigned toplines = 4;
   const unsigned ndimension = 3;
```
Use the template dimnsion.

kks32 commented on include/gmsh.tcc:
```diff
@@ -66,9 +64,13 @@ void GMSH::read_elements(const std::string& filename) {
   const unsigned toplines = 4;
   const unsigned ndimension = 3;
+  // specify element type 4 = tetrahedral
+  const unsigned element_type = 4;
```
Use an ENUM

ezrayst replied:
I will use `Tvertices` instead

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,23 +111,107 @@ void GMSH::read_elements(const std::string& filename) {
+  std::array<double, ncoordinates> verticesarray;
+
+  //! Iterate through element_
+  for (unsigned q = firstelement; q <= lastelement; ++q) {
```
I thought ezra's latest commit fixed this algorithm

kks32 replied:
```
+      //! Iterate through the 4 vertices to get coordinates
+      for (unsigned q = 0; q <= 3; ++q) {
+        //! Get the vertex wanted from the id 
+        verticesfind = vertices_.find(elementkeyvalues[q]);
+        //! For each vertex, store the coordinates
+        //! j = 0 -> [X], j = 0 -> [Y], j = 0 -> [Z]
+        for (unsigned j = 0; j <= 2; ++j) {
+          verticesarray[q * 3 + j] = verticesfind->second[j];
         }
       }
```

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,23 +111,107 @@ void GMSH::read_elements(const std::string& filename) {
+        }
+      }
+      // Centroid test
+      //++++++++++++++++++++++++++++++++++++++
```
Can you iterate here as well?

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,23 +111,107 @@ void GMSH::read_elements(const std::string& filename) {
+    if (coordinatesfind != elementcoordinates_.end()) {
+      //! Store coordinates in 3x4 matrix
+      Eigen::MatrixXd m(3, 4);
+      for (unsigned i = 0; i <= 3; ++i) {
```
Why do they have different indices 3 and 2?

cw646 committed:
:construction: Working on issue #14, #18, #19
Templatised dimension, variables and total length of coordinate array

kks32 commented on include/gmsh.h:
```diff
@@ -1,20 +1,21 @@
 #include <memory>
 #include <vector>
-
 // libraries
```
You don't need to explicity say local or libraries.

kks32 commented on include/gmsh.h:
```diff
@@ -1,20 +1,21 @@
+#include "point.h"
 
 //! \brief Generate Material Points from GMSH file
+template <unsigned Tdim, unsigned Tvert, unsigned Tncoords>
```
You don't need `Tncoords` use the product of Tdim * Tvertices.

kks32 replied:
Rename `Tvert` as `Tvertices`

kks32 commented on include/gmsh.h:
```diff
@@ -37,38 +38,40 @@ class GMSH : public Mesh {
 
  private:
-  //! Number of vertices
+  //! Number of vertices in total
```
//! Total number of vertices

kks32 commented on include/gmsh.tcc:
```diff
@@ -63,14 +65,11 @@ void GMSH::read_elements(const std::string& filename) {
 
   //! Array to store vertices coordinates
-  std::array<double, elementvertices> elementarray;
+  std::array<double, Tvert> elementarray;
```
Change `double` to `int` or `unsigned` for all ids

kks32 commented on include/gmsh.tcc:
```diff
@@ -213,7 +206,8 @@ void GMSH::compute_material_points() {
+template <unsigned Tdim, unsigned Tvert, unsigned Tncoords>
+void GMSH<Tdim, Tvert, Tncoords>::compute_stresses() {
 
   double density = 22;
```
These should be const

kks32 commented on src/main.cc:
```diff
@@ -16,7 +16,9 @@ int main(int argc, char** argv) {
-    std::unique_ptr<GMSH> mesh(new GMSH());
+    //! \Param[in] Tdim = Dimension, Tvert = n vertices in element, Tcoord =
+    //! Tdim*Tvert
+    std::unique_ptr<GMSH<3, 4, 12>> mesh(new GMSH<3, 4, 12>());
```
This should be a `Mesh` class object, not `GMSH`

Commit:
:art: fix some issues #21

Commit:
:construction: fixed more of #21

Commit:
:construction: clang-format

Commit:
:art: fix issue #20, make abstract class

kks32 commented on include/gmsh.tcc:
```diff
@@ -1,11 +1,12 @@
 //! Read vertex id and coordinates in GMSH
 //! \param[in] filename Input GMSH file
-void GMSH::read_vertices(const std::string& filename) {
+//! \Param[in] Tdim = Dimension, Tvert = n vertices in element
```
Please read the Doxygen style for variables.
`\tparam Tdim Dimension`

kks32 commented on include/gmsh.tcc:
```diff
@@ -28,29 +29,30 @@ void GMSH::read_vertices(const std::string& filename) {
 //! Read GMSH elements
 //! \param[in] filename Input GMSH file
-void GMSH::read_elements(const std::string& filename) {
+//! \Param[in] Tdim = Dimension, Tvert = n vertices in element
```
`\tparam`

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
         //! \brief Check element type
-        //! \details If element type not == to Tdim, skip element
-        if (elementtype != ndimension) {
+        //! \details If element type not == to specified Tvertices, skip
```
No `\details` inside the function!

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
+            << '\n';
+
+  //! Get the coordinates for each vertex of each element
+  GMSH::store_element_vertices();
```
use `this->`

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
+}
+
+//! \brief store each coordinates of an element
+//! \Param[in] Tdim = Dimension, Tvert = n vertices in element
```
Remove `=` and should be `\tparam`

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
+    if (elementfind != Mesh<Tdim, Tvertices>::elements_.end()) {
+
+      //! In each element, iterate to get vertices id's of the element
+      for (unsigned j = 0; j < 4; ++j) {
```
Can we not combine these two for loops? Shouldn't this be iterating to Tvertices?

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
+
+  unsigned arrayposition = 0;
+
+  typename std::map<unsigned, std::array<double, Tdim * Tvertices>>::iterator
```
Use `alias`

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
+  const unsigned lastelementcoord =
+      Mesh<Tdim, Tvertices>::elementcoordinates_.rbegin()->first;
+
+  for (unsigned t = firstelementcoord; t < lastelementcoord + 1; ++t) {
```
Why not use an iterator?

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
+  std::array<double, Tdim * Tvertices> verticesarray;
+
+  //! Iterate through element_
+  for (unsigned i = firstelement; i <= lastelement; ++i) {
```
Use an iterator.

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
-  std::cout << "Number of Elements: " << elements_.size() << '\n';
+      //! Centroid test
+      //++++++++++++++++++++++++++++++++++++++
+      for (unsigned i = 0; i < 3; ++i) {
```
i < Tdim, j < Tvertices

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
+  const double k0 = 0.5;
+  const double max_height = 3;
+  const double conv_factor = 10;
+  const unsigned stress_comp = 6;
```
This is just Tdim * 2, use Tdim * 2 instead.

kks32 commented on include/gmsh.tcc:
```diff
@@ -109,42 +110,142 @@ void GMSH::read_elements(const std::string& filename) {
   }
 
-  std::cout << "Compute initial stresses for material points.\n";
+  std::cout << "Initial stresses for material points computed.\n";
```
Remove cout

kks32 commented on include/io.tcc:
```diff
@@ -1,35 +1,36 @@
+    const std::vector<std::shared_ptr<Point<3>>>& materialpoints) {
 
   //! Output vertices file
   std::fstream vertices_file;
```
Change vertices to material points everywher.

kks32 commented on include/mesh.h:
```diff
@@ -3,19 +3,78 @@
 
  public:
-  //! Read vertices from mesh
+  //! \brief constructor for a mesh class
```
Remove these!

kks32 commented on include/mesh.h:
```diff
@@ -3,19 +3,78 @@
+  virtual unsigned nvertices() const { return 0; }
+
+  //! Return a vector of material points
+  virtual std::vector<std::shared_ptr<Point<Tdim>>> material_points() {
```
Why are these virtual?

kks32 commented on include/mesh.h:
```diff
@@ -3,19 +3,78 @@
+  }
+
+  //! Return a map of mesh element vertices
+  virtual std::map<unsigned, std::array<double, Tdim>> vertices() const {
```
Shouldn't be virtual

kks32 commented on include/mesh.h:
```diff
@@ -3,19 +3,78 @@
+  }
+
+  //! Return a map of element id & vertices id
+  virtual std::map<unsigned, std::array<double, Tvertices>> elements() const {
```
Not virutal

kks32 commented on include/mesh.h:
```diff
@@ -3,19 +3,78 @@
+  }
+
+  //! Return a vector of stresses
+  virtual std::vector<std::array<double, 6>> stress() const {
```
Not virutal

kks32 commented on include/mesh.h:
```diff
@@ -3,19 +3,78 @@
+  //! Total number of vertices
+  unsigned nvertices_;
+
+  std::vector<std::shared_ptr<Point<Tdim>>> materialpoints_;
```
Add comment

kks32 commented on include/mesh.h:
```diff
@@ -3,19 +3,78 @@
+  //! Stress vector in Voigt Notation
+  //! $\sigma_{xx}$ $\sigma_{yy}$ $\sigma_{zz}$
+  //! $\tau_{yz}$ $\tau_{zx}$ $\tau_{xy}$
+  std::vector<std::array<double, 6>> stress_;
```
Should be Tdim * 2

kks32 commented on include/gmsh.h:
```diff
@@ -21,34 +25,37 @@ class GMSH : public Mesh {
-  //! $\sigma_{xx}$ $\sigma_{yy}$ $\sigma_{zz}$ $\tau_{yz}$ $\tau_{zx}$ $\tau_{xy}$
-  std::vector<std::array<double, 6>> stress_;
+  //! Return a vector of stresses
+  std::vector<std::array<double, 6>> stress() const {
```
Should be defined in `Mesh`

kks32 replied:
Use the Mesh class's `stress_`.  using `Mesh<Tdim, Tvertices>::stress_`

kks32 commented on include/gmsh.h:
```diff
@@ -21,34 +25,37 @@ class GMSH : public Mesh {
-  //! Vector of vertices 
-  std::vector<std::shared_ptr<Point<3>>> vertices_;
+  //! Return a map of element id & vertices id
+  std::map<unsigned, std::array<double, Tvertices>> elements() const {
```
Should be defined in `Mesh`

kks32 commented on include/gmsh.h:
```diff
@@ -21,34 +25,37 @@ class GMSH : public Mesh {
-  //! Number of vertices
-  unsigned nvertices_;
+  //! Return a map of mesh element vertices
+  std::map<unsigned, std::array<double, Tdim>> vertices() const {
```
Should be defined in `Mesh`

kks32 commented on include/gmsh.h:
```diff
@@ -21,34 +25,37 @@ class GMSH : public Mesh {
-  //! Return a vector of elements
-  std::vector<std::shared_ptr<Point<3>>> elements() const { return elements_; }
+  //! Return a vector of material points
+  std::vector<std::shared_ptr<Point<Tdim>>> material_points() {
```
Should be defined in Mesh

kks32 commented on include/gmsh.h:
```diff
@@ -21,34 +25,37 @@ class GMSH : public Mesh {
-
-  //! Return a vector of vertices
-  std::vector<std::shared_ptr<Point<3>>> vertices() const { return vertices_; }
+  unsigned nvertices() const { return Mesh<Tdim, Tvertices>::nvertices_; }
```
Should be defined in Mesh

kks32 commented on src/main.cc:
```diff
@@ -16,14 +17,18 @@ int main(int argc, char** argv) {
 
     //! Mesh handler
-    std::unique_ptr<GMSH> mesh(new GMSH());
+    //! \Param[in] Tdim = Dimension, Tvert = n vertices in element
```
Why `\Param` in the middle of the main function?

kks32 commented on src/main.cc:
```diff
@@ -16,14 +17,18 @@ int main(int argc, char** argv) {
     //! Mesh handler
-    std::unique_ptr<GMSH> mesh(new GMSH());
+    //! \Param[in] Tdim = Dimension, Tvert = n vertices in element
+    GMSH<3, 4> gmsh;
```
GMSH should not appear!

kks32 commented on src/main.cc:
```diff
@@ -16,14 +17,18 @@ int main(int argc, char** argv) {
-    std::unique_ptr<GMSH> mesh(new GMSH());
+    //! \Param[in] Tdim = Dimension, Tvert = n vertices in element
+    GMSH<3, 4> gmsh;
+    Mesh<3, 4>* mesh = &gmsh;
```
Use a unique_ptr of base class `Mesh` but create it with a new `GMSH`

kks32 commented on src/main.cc:
```diff
@@ -16,14 +17,18 @@ int main(int argc, char** argv) {
-    //! Write vertices and stresses
-    io->write_vertices(mesh->vertices());
+    //! Write material points and stresses
+    io->write_vertices(mesh->material_points());
```
should be `write_material_points`

kks32 reviewed (changes requested):
Check inheritance and base class / derived class of Mesh and GMSH.

Commit:
:construction: changed doxygen  #21

kks32 commented on include/gmsh.tcc:
```diff
@@ -28,29 +29,30 @@ void GMSH::read_vertices(const std::string& filename) {
         istream >> vertex.at(0) >> vertex.at(1) >> vertex.at(2);
-
-        vertices_.emplace_back(new Point<ndimension>(vertid, vertex));
+        Mesh<Tdim, Tvertices>::vertices_.insert(std::make_pair(vertid, vertex));
```
Use `using Mesh<Tdim, Tvertices>::vertices_` in the header file, and you can use `vertices_`.

cw646 committed:
:construction: working on issues #14, #18, #19
Templatised Mesh class and IO to include dimension.
Fixed centroid test issues
GMSH now only uses Mesh material points container, not multiple variables.
Some cosmetic changes in IO.tcc

kks32 commented on include/gmsh.tcc:
```diff
@@ -1,6 +1,6 @@
 //! Read vertex id and coordinates in GMSH
+//! \tparam Tdim Dimension, Tvert Number of vertices in element
```
`Tvert` should be in the next line. Use `doxygen` to check if the comments are generated properly.

kks32 commented on include/gmsh.tcc:
```diff
@@ -49,8 +49,8 @@ void GMSH<Tdim, Tvertices>::read_vertices(const std::string& filename) {
 }
 
 //! Read GMSH elements
+//! \tparam Tdim Dimension, Tvert Number of vertices in element
```
Same `Tvert` in a new line

kks32 commented on include/gmsh.h:
```diff
@@ -21,34 +23,46 @@ class GMSH : public Mesh {
 
-  //! Return a vector of vertices
-  std::vector<std::shared_ptr<Point<3>>> vertices() const { return vertices_; }
+  //! Return a map of mesh element vertices
```
This should be in the `Mesh` class

kks32 commented on include/gmsh.h:
```diff
@@ -1,18 +1,20 @@
 //! \brief Generate Material Points from GMSH file
-class GMSH : public Mesh {
+template <unsigned Tdim, unsigned Tvertices>
+class GMSH : public Mesh<Tdim> {
```
Templatise the mesh with `Tvertices` as well.

kks32 commented on include/gmsh.h:
```diff
@@ -21,34 +23,46 @@ class GMSH : public Mesh {
-  //! Return a vector of elements
-  std::vector<std::shared_ptr<Point<3>>> elements() const { return elements_; }
+  //! Return a map of element id & vertices id
+  std::map<unsigned, std::array<unsigned, Tvertices>> elements() const {
```
This should go to the `Mesh` class

kks32 commented on include/gmsh.h:
```diff
@@ -21,34 +23,46 @@ class GMSH : public Mesh {
 
  private:
-  //! Number of vertices
+  //! Number of vertices in total
```
`//! Total number of vertices`

kks32 commented on include/gmsh.h:
```diff
@@ -21,34 +23,46 @@ class GMSH : public Mesh {
 
-  //! Vector of vertices 
-  std::vector<std::shared_ptr<Point<3>>> vertices_;
+  //! Stress vector in Voigt Notation
```
`stress_`, `vertices_`, and `elementcoords_` should be in base class `Mesh`.

kks32 commented on include/gmsh.h:
```diff
@@ -21,34 +23,46 @@ class GMSH : public Mesh {
-  std::vector<std::array<double, 6>> stress_;
+  //! Map to store element ID and vertices ID
+  std::map<double, std::array<double, Tvertices>> elements_;
+  std::map<double, std::array<double, Tdim * Tvertices>> elementcoordinates_;
```
Why coordinates of elements, is it not sufficient to map it to the ID of vertices?

kks32 commented on include/gmsh.tcc:
```diff
@@ -108,43 +109,130 @@ void GMSH::read_elements(const std::string& filename) {
+
+      //! In each element, iterate to get vertices id's of the element
+      for (unsigned j = 0; j < Tvertices; ++j) {
+        elementkeyvalues[j] = elementfind->second[j];
```
Use `.at` operator

kks32 commented on include/gmsh.tcc:
```diff
@@ -108,43 +109,130 @@ void GMSH::read_elements(const std::string& filename) {
+        elementkeyvalues[j] = elementfind->second[j];
+      }
+      //! Iterate through the vertices to get coordinates (4 for tetrahedral)
+      for (unsigned q = 0; q < Tvertices; ++q) {
```
Which `q`, defined in line 146 or 155?

kks32 commented on include/gmsh.tcc:
```diff
@@ -108,43 +109,130 @@ void GMSH::read_elements(const std::string& filename) {
+      //++++++++++++++++++++++++++++++++++++++
+      for (unsigned i = 0; i < Tdim; ++i) {
+        // reset array values
+        pointsarray[i] = 0;
```
Use STL algorithms or fill

kks32 replied:
Why not declare pointsarray here?

kks32 commented on include/gmsh.tcc:
```diff
@@ -108,43 +109,130 @@ void GMSH::read_elements(const std::string& filename) {
   //! Loop through the points to get vertical and horizontal stresses
   //! Note that tau (shear stress) is assumed 0
-  for (const auto& point : vertices_) {
+  for (const auto& point : Mesh<Tdim>::materialpoints_) {
```
Use `using` in the derived class to avoid these `Mesh<Tdim>::`

kks32 commented on include/io.tcc:
```diff
@@ -40,14 +46,17 @@ void IO::write_stresses(const std::vector<std::array<double, 6>>& stresses) {
     stress_file << stresses.size() << "\n";
 
     //! Stresses in Voigt Notation
+    // Stress needs ID too
```
OK!

kks32 commented on include/mesh.h:
```diff
@@ -1,21 +1,47 @@
   virtual void compute_stresses() = 0;
+
+  //! Return a vector of material points
+  std::vector<std::shared_ptr<Point<Tdim>>> write_material_points() {
```
rename to `material_points` as this function doesn't write anything.

kks32 commented on include/mesh.h:
```diff
@@ -1,21 +1,47 @@
+    return stress_;
+  }
+
+  //! Container for storing material points
```
Shouldn't these be in `protected`?

cw646 committed:
:construction: working on issues #14, #18, #19
Templatised Mesh class to include Tdim & Tvertices
Moved variables from GMSH to Mesh to avoid redeclaring in future
Auto detect for map itterators
Some cosmetic changes to function names

kks32 commented on include/gmsh.h:
```diff
@@ -14,7 +14,7 @@
 
```
Could you run `clang-format`, the includes are not in the right order.

cw646 committed:
:construction: working on issues #14, #18, #19
Clang format

kks32 commented on src/main.cc:
```diff
@@ -18,7 +18,7 @@ int main(int argc, char** argv) {
+    std::unique_ptr<Mesh<3, 4>> mesh(new GMSH<3, 4>());
 
     //! Read mesh and compute material points & stresses
     mesh->read_vertices(io->mesh_file_name());
```
Could we have a single function? `read_mesh` which calls both the vertices and elements?

kks32 committed:
:wrench: Fixes Eigen include in CI

kks32 committed:
:hammer: Fixes TravisCI missing eigen installation

kks32 committed:
:wrench: Clang has eigen in TravisCI

kks32 committed:
:wrench: Fix GCC5 deprecated declaration error

cw646 committed:
:art: read_mesh() replaces read_vertices() & read_elements() in Mesh;
In GMSH read_mesh() call's read_elements() once vertices are read in,
in read_mesh().

kks32 committed:
Merge branch 'develop' into points_within_element

kks32 committed:
:fire: Remove redundant variables and :art: Formatting

kks32 committed:
:computer: Using this pointer to call store elements

kks32 committed:
:fire: Remove FindEigen3.cmake in the root directory

kks32 committed:
:wrench: Fix read mesh calls to vertices and elements in GMSH
